traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go

```diff
@@ -1584,12 +1591,53 @@ func Validate(tx *sql.Tx, ds *tc.DeliveryServiceV5) error {
 	if err := validateTypeFields(tx, ds); err != nil {
 		errs = append(errs, fmt.Errorf("type fields: %w", err))
 	}
+	userErr, sysErr := validateRequiredCapabilities(tx, ds)
+	if userErr != nil {
+		errs = append(errs, errors.New("required capabilities: "+userErr.Error()))
+	}
+	if sysErr != nil {
+		errs = append(errs, errors.New("reading/ scanning required capabilities: "+sysErr.Error()))
+	}
 	if len(errs) == 0 {
 		return nil
 	}
 	return util.JoinErrs(errs)
 }
 
+func validateRequiredCapabilities(tx *sql.Tx, ds *tc.DeliveryServiceV5) (error, error) {
+	missing := make([]string, 0)
+	var missingCap string
+	query := `SELECT missing
+FROM (
+    SELECT UNNEST($1::TEXT[])
+    EXCEPT
+    SELECT UNNEST(ARRAY_AGG(name)) FROM server_capability) t(missing)
+`
+	if len(ds.RequiredCapabilities) > 0 {
+		rows, err := tx.Query(query, pq.Array(ds.RequiredCapabilities))
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			err = rows.Scan(&missingCap)
+			if err != nil {
+				return nil, err
+			}
+			missing = append(missing, missingCap)
+		}
+		if len(missing) > 0 {
+			var msg string
+			for _, m := range missing {
+				msg = msg + m + " "
+			}
+			userErr := fmt.Errorf("the following capabilities do not exist: %s", msg)
```

Comment:
there's a standard library function for this: [`strings.Join`](https://pkg.go.dev/strings#Join)